### PR TITLE
Reduce allocations in some `Array.Parallel` funcs

### PR DESCRIPTION
**Pull request by brianrourkeboll**
## Description

Followup to #14827.

- Avoid extra `'T option` allocations in `Array.Parallel.tryFindIndex`, `Array.Parallel.tryFind`, and `Array.Parallel.tryPick`.

## Checklist

- [ ] Test cases added.
- [ ] Performance benchmarks added in case of performance changes.
- [ ] Release notes entry updated.

### CODE REVIEW

**Comment by brianrourkeboll** [comment]
Hmm, a3a50c0 reduces allocations significantly, but while it is faster in many scenarios, it is slower in others, depending on the size of the array and the predicate. I may take that change out, unless we know which scenarios we'd prefer to optimize for.

**Comment by brianrourkeboll** [comment]
> Hmm, [a3a50c0](https://github.com/dotnet/fsharp/commit/a3a50c0c6525a122e0e468c69d20f2402b3e21c8) reduces allocations significantly, but while it is faster in many scenarios, it is slower in others, depending on the size of the array and the predicate. I may take that change out, unless we know which scenarios we'd prefer to optimize for.

Removed that for now.

**Review by psfinaki — APPROVED** [review]
Good stuff, thank you!

File: src/FSharp.Core/array.fs [hunk 1]
```diff
@@ -2179,9 +2179,8 @@ module Array =
             // Not exists $condition <==> (opposite of $condition is true forall)
             exists (predicate >> not) array |> not
 
-        [<CompiledName("TryFindIndex")>]
-        let tryFindIndex predicate (array: _ array) =
-            checkNonNull "array" array
+        let inline tryFindIndexAux predicate (array: _ array) =
```
Comment (vzarytovskii) [inline src/FSharp.Core/array.fs:2182]: Wondering if we want to inline predicate as well, and if it's worth it.

**Review by vzarytovskii — APPROVED** [review]

File: src/FSharp.Core/array.fs (hunk 1 above)
Comment (brianrourkeboll) [reply src/FSharp.Core/array.fs:2182]: My guess would be no, unless the top-level function were also marked `inline`.

Comment (vzarytovskii) [reply src/FSharp.Core/array.fs:2182]: Yeah, I meant both of them, on public surface as well as on auxiliary function.

Comment (brianrourkeboll) [reply src/FSharp.Core/array.fs:2182]: I would run some benchmarks, but I'm away from my computer for a week or two

**Review by KevinRansom — CHANGES_REQUESTED** [review]
Looks good, however it needs the release notes updated.

File: src/FSharp.Core/array.fs (hunk 1 above)
Comment (vzarytovskii) [reply src/FSharp.Core/array.fs:2182]: Yeah, it's not that important, we can always do some additional improvements later

**Pull request merged**